### ISSUE TITLE
 [Prusti internal error] generating fold-unfold Viper statements failed 

### ISSUE THREAD

**Issue by jmc-figueira** [viperproject/prusti-dev#471]
Similarly to #447 (although with a different cause, I believe), Prusti reports an internal error in the following code:

```Rust
#[ensures(result == true ==> self.add_set.contains(&val))]
#[ensures(result == true ==> !self.remove_set.contains(&val))]
#[ensures(result == true ==> self.length() == old(self.length()) + 1)]
#[ensures(result == false ==> self.length() == old(self.length()))]
fn add(&mut self, val: u64) -> bool {
    if self.remove_set.contains(&val) {
        false
    }
    else {
        self.add_set.insert(val)
    }
}

#[ensures(result == true ==> self.remove_set.contains(&val))]
#[ensures(result == true ==> !self.add_set.contains(&val))]
#[ensures(result == true ==> self.length() == old(self.length()) + 1)]
#[ensures(result == false ==> self.length() == old(self.length()))]
fn remove(&mut self, val: u64) -> bool {
    if !self.add_set.contains(&val) {
        false
    }
    else {
        self.remove_set.insert(val)
    }
}

#[pure]
#[ensures(result ==> (self.add_set.contains(&val) && !self.remove_set.contains(&val)))]
#[ensures(!result ==> (!self.add_set.contains(&val) || self.remove_set.contains(&val)))]
fn contains(&self, val: u64) -> bool {
    self.add_set.contains(&val) && !self.remove_set.contains(&val)
}
```

More information about the error:

```Rust
error: [Prusti internal error] generating fold-unfold Viper statements failed (RequiresFolding("u64", [Local(_2: Ref(u64), Position { line: 36, column: 34, id: 169 })], Write, None, Position { line: 36, column: 34, id: 169 }))
  --> src/lib.rs:41:5
   |
41 | /     fn add(&mut self, val: u64) -> bool {
42 | |         if self.remove_set.contains(&val) {
43 | |             false
44 | |         }
...  |
47 | |         }
48 | |     }
   | |_____^

error: [Prusti internal error] generating fold-unfold Viper statements failed (RequiresFolding("u64", [Local(_2: Ref(u64), Position { line: 50, column: 34, id: 320 })], Write, None, Position { line: 50, column: 34, id: 320 }))
  --> src/lib.rs:54:5
   |
54 | /     fn remove(&mut self, val: u64) -> bool {
55 | |         if !self.add_set.contains(&val) {
56 | |             false
57 | |         }
...  |
60 | |         }
61 | |     }
   | |_____^

error: [Prusti internal error] generating unfolding Viper expressions failed (FailedToObtain(Pred(Local(_2: Ref(u64), Position { line: 70, column: 27, id: 351 }), read)))
  --> src/lib.rs:72:5
   |
72 | /     fn contains(&self, val: u64) -> bool {
73 | |         self.add_set.contains(&val) && !self.remove_set.contains(&val)
74 | |     }
   | |_____^
```

[Full code for this example](https://gist.github.com/jmc-figueira/9b0d5229662e15ca8358cdf6196eec90)
[Viper program dump](https://gist.github.com/jmc-figueira/b255fa700841aecc1d7372562c8f4032)

**Comment by vakaras** [comment]
Thank you for the bug report! I suspect that the problem is generics, but I would need to investigate deeper (the current implementation of generics is completely broken, we are planning to reimplement them once we have a chance). In that case, a possible workaround should be to use a wrapper:
```rust
struct HashSetu64 {
  internal: HashSet<u64>,
}
impl HashSetu64 {
    #[trusted]  // Tell Prusti to ignore the function body.
    #[ensures(result.is_empty())]
    pub fn new() -> Self {
        Self { internal: HashSet::new() }
    }
```

**Comment by jmc-figueira** [comment]
> Thank you for the bug report! I suspect that the problem is generics, but I would need to investigate deeper (the current implementation of generics is completely broken, we are planning to reimplement them once we have a chance). In that case, a possible workaround should be to use a wrapper:
> 
> ```rust
> struct HashSetu64 {
>   internal: HashSet<u64>,
> }
> impl HashSetu64 {
>     #[trusted]  // Tell Prusti to ignore the function body.
>     #[ensures(result.is_empty())]
>     pub fn new() -> Self {
>         Self { internal: HashSet::new() }
>     }
> ```

Thank you for the advice! It did help get more conditions verified, (insert and remove produce no errors now), but still, some issues remain:

```Rust
error: [Prusti internal error] generating fold-unfold Viper statements failed (FailedToObtain(Pred(Field(AddrOf(Local(_5: Ref(ref$m_HashSetU64$_beg_$_end_), Position { line: 0, column: 0, id: 0 }), TypedRef("ref$m_HashSetU64$_beg_$_end_"), Position { line: 68, column: 56, id: 72 }), val_ref: Ref(m_HashSetU64$_beg_$_end_), Position { line: 0, column: 0, id: 0 }), read)))
  --> src/lib.rs:65:5
   |
65 | /     fn merge(&self, other: &TwoPSet) -> TwoPSet {
66 | |         let mut result = TwoPSet::new();
67 | |
68 | |         result.add_set = self.add_set.union(&other.add_set);
...  |
71 | |         result
72 | |     }
   | |_____^

error: [Prusti internal error] generating unfolding Viper expressions failed (FailedToObtain(Pred(Local(_2: Ref(u64), Position { line: 54, column: 27, id: 568 }), read)))
  --> src/lib.rs:56:5
   |
56 | /     fn contains(&self, val: u64) -> bool {
57 | |         self.add_set.contains(&val) && !self.remove_set.contains(&val)
58 | |     }
   | |_____^
```

[Relevant updated code](https://gist.github.com/jmc-figueira/a2c34a3874d1b1617e14b1659f2dd9a4)

**Comment by fpoli** [comment]
PR #473 fixes two bugs, after which we are left with #472 and #474.

A workaround for #474 is to change the type of `val` in `contains` to be `&u64`, such that no shared reference is created within the pure function:
```rust
    #[pure]
    #[ensures(result ==> (self.add_set.contains(&val) && !self.remove_set.contains(&val)))]
    #[ensures(!result ==> (!self.add_set.contains(&val) || self.remove_set.contains(&val)))]
    fn contains(&self, val: &u64) -> bool {
        self.add_set.contains(val) && !self.remove_set.contains(val)
    }
```

I'll close this issue with PR #473 as #474 is more specific.

**Comment by jmc-figueira** [comment]
Thank you for your quick fix!
